docs/reference/search/semantic-search.asciidoc

```diff
@@ -0,0 +1,133 @@
+[[knn-search-api]]
+=== Semantic search API
+++++
+<titleabbrev>Semantic search</titleabbrev>
+++++
+
+experimental::[]
+Semantic search uses a text embedding NLP model to generate a dense vector from the input query string.
+The resulting dense vector is then used in a <Knn Search> against an index containing dense vectors
+created with the same text embedding model. The search results are semantically similar as learned
+by the model.
+
+////
+[source,console]
+----
+PUT my-index
+{
+  "mappings": {
+    "properties": {
+      "text_embedding": {
+        "type": "dense_vector",
+        "dims": 512,
+        "index": true,
+        "similarity": "cosine"
+      }
+    }
+  }
+}
+----
+////
+
+[source,console]
+----
+GET my-index/_semantic_search
+{
+  "query_string": "A picture of a snow capped mountain",
+  "model_id": "my-text-embedding-model",
+  "knn": {
+    "field": "text_embedding",
+    "k": 10,
+    "num_candidates": 100
+  }
+}
+----
+// TEST[skip:TBD]
+
+
+[[semantic-search-api-request]]
+==== {api-request-title}
+
+`GET <target>/_semantic_search`
+
+`POST <target>/_semantic_search`
+
+[[semantic-search-api-prereqs]]
+==== {api-prereq-title}
+
+* If the {es} {security-features} are enabled, you must have the `read`
+<<privileges-list-indices,index privilege>> for the target data stream, index,
+or alias.
+
+[[Semantic-search-api-desc]]
+==== {api-description-title}
+
+The Semantic search API uses a text embedding model to create a dense vector
+representation of the query string.
+
+
+[[Semantic-search-api-path-params]]
+==== {api-path-parms-title}
+
+`<target>`::
+(Optional, string) Comma-separated list of data streams, indices, and aliases
+to search. Supports wildcards (`*`). To search all data streams and indices,
+use `*` or `_all`.
+
+[role="child_attributes"]
+[[Semantic-search-api-query-params]]
+==== {api-query-parms-title}
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=routing]
+
+[role="child_attributes"]
+[[semantic-search-api-request-body]]
+==== {api-request-body-title}
+
+`model_id`::
+(Required, string)
+include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=model-id]
+
+`query_string`::
```

Comment:
I'm happy to leave this as it is for this PR as the new endpoint is experimental, but `query_string` might not be the best field name here.

The existing `query_string` query supports an advanced search syntax: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html

What is being supplied here is more akin to what would be supplied to a `match` query: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html

So maybe something like `match_text` would be more in keeping with prior art.  Or alternatively, since kNN search used `query_vector`, maybe `query_text` so that it's closer to that but doesn't clash with the existing `query_string` query's name.

But, like I say, I am happy to get this merged so that people can start playing with it, then revisit naming before GA.